connect/runtime/src/main/java/org/apache/kafka/connect/runtime/PluginDiscovery.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.kafka.connect.tools.VerifiableSinkConnector;
+import org.apache.kafka.connect.tools.VerifiableSourceConnector;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.apache.kafka.connect.util.ReflectionsUtil;
+import org.reflections.Reflections;
+import org.reflections.util.ClasspathHelper;
+import org.reflections.util.ConfigurationBuilder;
+
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+
+public class PluginDiscovery {
+
+    private static final List<Class<? extends Connector>> CONNECTOR_EXCLUDES = Arrays.asList(
+            VerifiableSourceConnector.class, VerifiableSinkConnector.class,
+            MockConnector.class, MockSourceConnector.class, MockSinkConnector.class,
+            SchemaSourceConnector.class
+    );
+
+    private static final List<Class<? extends Transformation>> TRANSFORMATION_EXCLUDES = Arrays.asList();
+
+    private static boolean scanned = false;
+    private static List<ConnectorPluginInfo> validConnectorPlugins;
+    private static List<Class<? extends Transformation>> validTransformationPlugins;
+
+    public static synchronized List<ConnectorPluginInfo> connectorPlugins() {
+        scanClasspathForPlugins();
+        return validConnectorPlugins;
+    }
+
+    public static synchronized List<Class<? extends Transformation>> transformationPlugins() {
+        scanClasspathForPlugins();
+        return validTransformationPlugins;
+    }
+
+    public static synchronized void scanClasspathForPlugins() {
+        if (scanned) return;
+        ReflectionsUtil.registerUrlTypes();
+        final Reflections reflections = new Reflections(new ConfigurationBuilder().setUrls(ClasspathHelper.forJavaClassPath()));
+        validConnectorPlugins = Collections.unmodifiableList(connectorPlugins(reflections));
+        validTransformationPlugins = Collections.unmodifiableList(transformationPlugins(reflections));
+        scanned = true;
+    }
+
+    private static List<ConnectorPluginInfo> connectorPlugins(Reflections reflections) {
+        final Set<Class<? extends Connector>> connectorClasses = reflections.getSubTypesOf(Connector.class);
+        connectorClasses.removeAll(CONNECTOR_EXCLUDES);
+
+        final List<ConnectorPluginInfo> connectorPlugins = new ArrayList<>(connectorClasses.size());
+        for (Class<? extends Connector> connectorClass : connectorClasses) {
+            if (isConcrete(connectorClass)) {
+                connectorPlugins.add(new ConnectorPluginInfo(connectorClass.getCanonicalName()));
+            }
+        }
+
+        Collections.sort(connectorPlugins, new Comparator<ConnectorPluginInfo>() {
+            @Override
+            public int compare(ConnectorPluginInfo a, ConnectorPluginInfo b) {
+                return a.clazz().compareTo(b.clazz());
```

Comment:
The Transformations are sorted by comparing the canonical names of the classes. Should we do that here as well?